File: Golang/ex12.go
package main

import "fmt"

type Produto struct {
	nome  string
	preco float64
}

func (p Produto) Somar(outro Produto) Produto {
	return Produto{
		nome:  "Soma de " + p.nome + " e " + outro.nome,
		preco: p.preco + outro.preco,
	}
}

func main() {
	produto1 := Produto{"Nescau", 50.0}
	produto2 := Produto{"Toddy", 30.0}

	produtoSoma := produto1.Somar(produto2)
	fmt.Printf("%s: R$%.2f\n", produtoSoma.nome, produtoSoma.preco)
}

File: Java/ex12.java
class Produto {
    private String nome;
    private double preco;

    public Produto(String nome, double preco) {
        this.nome = nome;
        this.preco = preco;
    }

    public Produto somar(Produto outro) {
        return new Produto("Soma de " + this.nome + " e " + outro.nome, this.preco + outro.preco);
    }

    @Override
    public String toString() {
        return String.format("%s: R$%.2f", nome, preco);
    }

    public static void main(String[] args) {
        Produto produto1 = new Produto("Café", 50.0);
        Produto produto2 = new Produto("Energético", 30.0);

        Produto produtoSoma = produto1.somar(produto2);
        System.out.println(produtoSoma);
    }
}
